Update README

stephencrowell opened the pull request:
Solve #5

stephencrowell committed:
Initial update to README

stephencrowell committed:
Update json format

stephencrowell committed:
Remove periods on numbered lists

stephencrowell committed:
Fix typos

stephencrowell commented:
@ebrahimebrahim I have the initial structure and info down. I don't think I missed anything major. Feel free to take a look when you get a chance.

ebrahimebrahim commented:
awesome, thank you! I should get a chance to look it over and help out tomorrow, hopefully

ebrahimebrahim commented:
I am planning to push some edits directly into your branch update-readme

(please let me know if you prefer I not do that and instead leave review comments only)

once done I will let you know the ball is back in your court

stephencrowell commented:
> I am planning to push some edits directly into your branch update-readme
> 
> (please let me know if you prefer I not do that and instead leave review comments only)
> 
> once done I will let you know the ball is back in your court

Go ahead and push your changes whene

ebrahimebrahim committed:
add first pass of ebrahim revisions to readme

stephencrowell committed:
Fix formatting

ebrahimebrahim committed:
Update README.md

add comment on limitations

ebrahimebrahim commented on README.md:
````diff
@@ -1,7 +1,133 @@
+2. Clone this project.
+    ```sh
+    git clone git@github.com:KitwareMedical/lungair-fhir-server.git
+    ```
````
An additional setup step we need here is _setting up the python environment_.

@stephencrowell I recommend:
- creating a venv (with python version 3.9 or earlier)
- pip installing to it _only the things that seem necessary_ for this project
- activating the venv and testing this project to make sure it runs
- `pip freeze > requirements.txt`
- commit the requirements file to this branch
- add `pip install -r requirements.txt` to the instructions here

ebrahimebrahim commented on README.md:
```diff
@@ -1,7 +1,133 @@
+4. Implement `create_patient` and `get_patient_observations` for
+your `PatientDataSource` implementation. Read the doc strings in
+`patient_data_source.py` for more information.
+5. Implement `get_observation_type` and `get_value` for your
```
Looking at this, I'm starting to think an example would be a lot more useful than a step-by-step description like this.

I made up some example data: [example.csv](https://github.com/KitwareMedical/lungair-fhir-server/files/9823718/example.csv)
We can show the table in markdown form, like "suppose you have this data and you want to populate a fhir server with it"

Let's add the observation type for body weight:
```json
"bodyweight":
{
    "display_string": "body weight",
    "unit_code": "kg",
    "loinc_code": "29463-7"
}
```
(see [here](https://loinc.org/29463-7/) for the loinc code description)


and then instead of step-by-step instructions we can just show how you would subclass the appropriate classes, load the data with pandas, implement the correct methods, make a json, etc.

What do you think?

stephencrowell replied:
I am a bit confused here. Are we adding body weight observation type as a type we implement or is this just for the example?

ebrahimebrahim replied:
my intention was to add it to our observation types json so that we support it,  might as well

stephencrowell replied:
In either case, I think it would be best to add another file showing a walk-through of all this information. At the end of the basic steps, we can say a more in depth explanation/walk-through can be found here. I think having an in-depth explanation of how to add a data source inside the README will make it cluttered.

We could even change the adding custom data sources to say something along the lines of "It is possible to add a custom data source by creating implementations of `Observation` and `PatientDataSource` from `patient_data_source.py`].  A more in-depth explanation can be found here." This will allow us to go into as much detail as we want inside the walk-through without extending the length of the README to a ridiculous size.

ebrahimebrahim replied:
Hmm sure we can separate it out. I like that!

How about we make a directory in this repo called "example" and put a README.md in there, example.csv that I just shared, and also put in `example_data_source.py` which defines `ExampleDataSource` and `ExampleObservation` and `ExamplePatient`, and finally `example_data_source.json`.

Then we can link to that folder here

stephencrowell committed:
Add package requirements

stephencrowell committed:
Addbodyweight as an observation type

stephencrowell committed:
Add example for custom observation

stephencrowell committed:
Fix doc string to encourage implementation

stephencrowell committed:
Change import to not require files to be in data_sources

stephencrowell committed:
Fix typos and ensured example works

ebrahimebrahim committed:
Rename example.md to README.md

this makes github display it on the page of the example folder while navigating the source tree

ebrahimebrahim commented on populate_fhir_server.py:
```diff
@@ -12,7 +12,7 @@
 with open(args.json_file) as json_file: # Import data source and create instance of data source
   data_source_dict = json.load(json_file)
-  exec('from data_sources.{0} import {1}'.format(data_source_dict['module_name'], data_source_dict['class_name']))
+  exec('from {0} import {1}'.format(data_source_dict['module_name'], data_source_dict['class_name']))
```
This is a nice improvement but now I'm thinking we can do better.
Here we still need to specify a valid import string in our json, like "`data_sources.mimic3`, which is not very friendly. For example if I want `MyDataSource.py` to live in some arbitrary place on my disk then I will not be able to import it.

Here's something we can try: replace the json argument `module_name` by `module_path` and make it be a _path to the module `.py` file that contains the user's data source implementation_. Then I think you can import from the path using this tool: https://docs.python.org/3/library/importlib.html#importlib.util.spec_from_file_location

If we do it this way can we also do away with the empty `__init__.py` in `example/` and `data_sources/`?

ebrahimebrahim commented on example/example_data_source.py:
```diff
@@ -0,0 +1,45 @@
+from collections.abc import Iterable
```
Hmm I'm on the fence here... do you think we should remove this `Iterable` import and remove the type annotations from this file? This is an example, not an API, so I'm thinking it should look as uncluttered as possible. Someone taking a look at this should think "oh that looks simple, I can probably make something like that". Not everyone knows about type annotations so I'm worried they hurt us in this case.

Then again sometimes the type annotations make it clearer what's going on, especially on function arguments when you're not seeing how the function is going to be called.

Maybe a good middle ground is to keep type annotations on function arguments, but get rid of them for return values? Or is that weird...

opinion wanted

ebrahimebrahim committed:
Update README.md

ebrahimebrahim commented on data_sources/patient_data_source.py:
```diff
@@ -21,7 +21,7 @@ def get_gender(self) -> Gender:
 		"""Return a custom identifier for the Patient. This is not the ID that will be used internally by the server.
 		This could be used to link the Patient to its point of origin in the data source, for example. Implementation
-		is not required."""
+		is not required, but highly recommend for implementing get_patient_observations"""
```
"but highly recommend for" --> "but may be needed for"

(thinking about it with the example we are working on, I think there are normal situations where it would not be needed)

ebrahimebrahim commented on example/example_data_source.py:
```diff
@@ -0,0 +1,45 @@
+    mask1 = self.data['patient_id'].duplicated(keep = 'first') # Get first occurance of patient_id
+    mask2 = self.data['patient_id'].duplicated(keep = False) # Mark duplicate patient_ids
+    mask = ~mask1 | ~mask2
+    return (ExamplePatient(row) for _, row in self.data[mask].iterrows())
```
this can be shortened and made more readable:
```py
unique_patients = self.data.drop_duplicates('patient_id')
return (ExamplePatient(row) for _, row in unique_patients.iterrows())
```

update example readme.md also

ebrahimebrahim commented on example/example_data_source.py:
```diff
@@ -0,0 +1,45 @@
+    return (ExamplePatient(row) for _, row in self.data[mask].iterrows())
+
+  def get_patient_observations(self, patient : Patient) -> Iterable[Observation]:
+    return (ExampleObservation(row) for _, row in self.data[self.data['patient_id'] == int(patient.get_indentifier_value())].iterrows())
```
readability improvement:

```py
patient_id = int(patient.get_indentifier_value())
observations_for_patient = self.data[self.data['patient_id']==patient_id]
return (ExampleObservation(row) for _, row in observations_for_patient.iterrows())
```

similarly can update the example readme

ebrahimebrahim committed:
Update README.md

ebrahimebrahim commented on example/README.md:
```diff
@@ -0,0 +1,200 @@
+## Creating JSON config file
+
+In order for `populate_fhir_server.py` to recognize `ExampleDataSource`, we need to create
+a new JSON config file with the proper information. [example.json](example.json) shows the format
```
Once the json args like `module_path`, `class_name` are finalized in how they should work (see my other comment on using `module_path`), I think we shouldn't just link to example.json here, but rather we should show the contents and explain `args`, `module_path`, `class_naem`

ebrahimebrahim reviewed (changes requested):
One writing inconsistency I'm seeing is a mixture of "you" and "we".

I think "we" sounds a lot better, so let's switch to using "we" throughout.

example:
> After following these steps, you will be able to run `populate_fhir_server.py` with your JSON file as an argument.

becomes

> After following these steps, **we** will be able to run `populate_fhir_server.py` with **our** JSON file as an argument.

stephencrowell committed:
Update import to use importlib

stephencrowell committed:
Remove type annontations in implemented classes

stephencrowell committed:
Update READMEs and doc strings

stephencrowell committed:
Update example README

stephencrowell commented:
@ebrahimebrahim ready for review again. There are still an occasional cases of "you" in the READMEs, but they are specific to changes the user will make (port on docker and implementation of more methods in their implementation of PDS). Changing those to "we" did not make much sense IMO.

ebrahimebrahim committed:
Update README.md

Getting rid of some of the "we" in the readme where it doesn't feel quite right.

ebrahimebrahim committed:
Update README.md

improved phrasing, hopefully

ebrahimebrahim commented on example/README.md:
```diff
@@ -0,0 +1,214 @@
+- `args` are the arguements passed into the `PatientDataSource.__init__` implementation.
+- `class_name` is the name of the `PatientDataSource` subclass.
+- `module_path` is the path to the `PatientDataSource` implementation.
+- `module_name` is the name of the python file of the `PatientDataSource` implementation.
```
Is `module_name` still needed in the json config now that we have `module_path`? Can `module_name` be deduced from `module_path`?

stephencrowell replied:
It can be. I'll make the changes now.

ebrahimebrahim reviewed (commented):
Almost ready, I made a few phrasing changes as well

Has everything been tested tested since the changes that use importlib?

stephencrowell commented:
> Has everything been tested tested since the changes that use importlib?

I ran all 3 PDS implementations and had no issues.

stephencrowell committed:
Remove module_name
